smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/IncomingKafkaBatchRecords.java

```diff
@@ -0,0 +1,106 @@
+            KafkaFailureHandler onNack, boolean cloudEventEnabled, boolean tracingEnabled) {
+        List<IncomingKafkaRecord<K, T>> incomingRecords = new ArrayList<>();
+        Map<TopicPartition, IncomingKafkaRecord<K, T>> latestOffsetRecords = new HashMap<>();
+        for (TopicPartition partition : records.partitions()) {
+            IncomingKafkaRecord<K, T> latestOffsetRecord = null;
+            for (ConsumerRecord<K, T> record : records.records(partition)) {
+                IncomingKafkaRecord<K, T> rec = new IncomingKafkaRecord<>(record, commitHandler, onNack,
+                        cloudEventEnabled, tracingEnabled);
+                incomingRecords.add(rec);
+                latestOffsetRecord = latestOffsetRecords.compute(partition,
+                        (tp, cr) -> (cr == null || rec.getOffset() > cr.getOffset()) ? rec : cr);
+            }
+            if (latestOffsetRecord != null) {
+                commitHandler.received(latestOffsetRecord);
+            }
+        }
+        this.incomingRecords = Collections.unmodifiableList(incomingRecords);
+        this.latestOffsetRecords = Collections.unmodifiableMap(latestOffsetRecords);
+        this.metadata = Metadata.of(new IncomingKafkaBatchRecordsMetadata<>(records));
+    }
+
+    @Override
+    public List<T> getPayload() {
+        return this.incomingRecords.stream().map(KafkaRecord::getPayload).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<KafkaRecord<K, T>> getRecords() {
+        return this.incomingRecords;
+    }
+
+    @Override
+    public Iterator<KafkaRecord<K, T>> iterator() {
+        return this.getRecords().iterator();
+    }
+
+    @Override
+    public Map<TopicPartition, KafkaRecord<K, T>> getLatestOffsetRecords() {
+        return this.latestOffsetRecords;
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public Supplier<CompletionStage<Void>> getAck() {
+        return this::ack;
+    }
+
+    @Override
+    public Function<Throwable, CompletionStage<Void>> getNack() {
+        return this::nack;
+    }
+
+    @Override
+    public CompletionStage<Void> ack() {
+        return Uni.combine().all().unis(this.latestOffsetRecords.values().stream()
+                .map(record -> Uni.createFrom().completionStage(record.ack()))
+                .collect(Collectors.toList()))
+                .collectFailures()
+                .discardItems()
+                .convert().toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Void> nack(Throwable reason, Metadata metadata) {
+        return Uni.combine().all().unis(this.incomingRecords.stream()
+                .map(record -> Uni.createFrom().completionStage(record.nack(reason, metadata)))
+                .collect(Collectors.toList()))
+                .collectFailures()
+                .discardItems()
+                .convert().toCompletionStage();
+    }
```

Comment:
For `ack`, we only use the latest record for each partition, so that's fine, but here, we `nack` each record, so maybe order matters? (Not sure how `Uni.combine().all(...)` handles ordering, just thinking out loud.)

Reply:
I asked myself the same question. But I assume that the zip operation will execute sequentially on `Uni`.

Reply:
Took another look at this. `Uni.combine.all(...)` isn't really a zip implementation. What we need here is `Multi.createBy().concatenating().streams(...)`.

Reply:
I changed both `nack` and `ack` to Multi concatenate for the sake of simplicity.

Reply:
+1, I think this makes sense now.